File: Topic_2_task_1/solution_1.py
def get_possible_sums(coins: list[int]) -> set[int]:
    max_sum = sum(coins)
    dp = [False] * (max_sum + 1)
    dp[0] = True
    for coin in coins:
        for j in range(max_sum, coin - 1, -1):
            if dp[j - coin]:
                dp[j] = True
    return {s for s in range(max_sum + 1) if dp[s]}

def count_max_cost_1(buyer: list[int], seller: list[int]) -> int | None:
    buyer_sums = get_possible_sums(buyer)
    seller_sums = get_possible_sums(seller)
    
    possible_p = set()
    for b in buyer_sums:
        for s in seller_sums:
            p = b - s
            if p > 0:
                possible_p.add(p)
    
    max_affordable = max(buyer_sums)
    for p in range(max_affordable, 0, -1):
        if p not in possible_p:
            return p
    
    if max_affordable == 0:  
        print ("У покупателя нет денег!")
        return None
    
    if len(seller) == 0:
        max_not_affordable = max_affordable - 1
        min_buyer_sum = min(buyer_sums)
        print(buyer_sums)
        print(max_not_affordable)
        while max_not_affordable >= min_buyer_sum:
            if buyer_sums.__contains__(max_not_affordable):
                max_not_affordable -= 1
            else:
                return max_not_affordable
        print("Покупатель может точно оплатить любую стоимость до своей максимальной суммы")  
        return None
    else:
        if min(buyer) <= min(seller):
            print("Покупатель может точно оплатить любую стоимость до своей максимальной суммы")
            return None
        return min(buyer) - min(seller) - 1

File: AaDS_project/AaDS_project/Topic_2_task_1/Solution_1.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace AaDS_project.Topic_2_task_1;

public class Solution_1
{
    public static HashSet<int> GetPossibleSums(List<int> coins)
    {
        int maxSum = coins.Sum();
        bool[] dp = new bool[maxSum + 1];
        dp[0] = true;

        foreach (int coin in coins)
        {
            for (int j = maxSum; j >= coin; j--)
            {
                if (dp[j - coin])
                {
                    dp[j] = true;
                }
            }
        }

        HashSet<int> result = new();
        for (int i = 0; i <= maxSum; i++)
        {
            if (dp[i])
                result.Add(i);
        }

        return result;
    }

    public static int? CountMaxCost(List<int> buyer, List<int> seller)
    {
        var buyerSums = GetPossibleSums(buyer);
        var sellerSums = GetPossibleSums(seller);

        HashSet<int> possibleP = new();

        foreach (int b in buyerSums)
        {
            foreach (int s in sellerSums)
            {
                int p = b - s;
                if (p > 0)
                {
                    possibleP.Add(p);
                }
            }
        }

        int maxAffordable = buyerSums.Max();

        for (int p = maxAffordable; p > 0; p--)
        {
            if (!possibleP.Contains(p))
            {
                return p;
            }
        }

        if (maxAffordable == 0)
        {
            Console.WriteLine("У покупателя нет денег!");
            return null;
        }

        if (buyer.Min() <= seller.Min())
        {
            Console.WriteLine("Покупатель может точно оплатить любую стоимость до своей максимальной суммы");
            return null;
        }

        return buyer.Min() - seller.Min() - 1;
    }
}
